[GeoMechanicsApplication] Clean up UDSM-based constitutive laws

avdg81 opened the pull request:
**📝 Description**
Cleaned up and generalized the UDSM-based constitutive laws.

**🆕 Changelog**
- The UDSM-based constitutive laws are no longer copy-able, but they are move-able. Use member function `Clone` for creating deep copies. Compile-time checks have been added to enforce this part of the design.
- By introducing a data member of type `ConstitutiveLawDimension`, we have been able to generalize class `SmallStrainUDSM3DLaw` such that class `SmallStrainUDSM2DPlaneStrainLaw` became redundant. The new data member is used for getting the strain size as well as the local space dimension. Consequently, we could get rid of static data members `VoigtSize` and `Dimension`.
- Class `SmallStrainUDSM3DLaw` has been renamed to `SmallStrainUDSMLaw`. The corresponding files have been renamed, too. Class `SmallStrainUDSM2DPlaneStrainLaw` has been removed.
- Moved function definitions from header files to the corresponding implementation files.
- Removed a few `override`s that were identical to the base class implementation.
- Removed many redundant comments.
- Removed several redundant name qualifiers.
- Removed several unused member functions.
- Removed several `KRATOS_TRY` and `KRATOS_CATCH` constructs, to avoid catching errors in places where they can't be properly handled anyway.
- Made the implementation of member `Clone` complete (several data members were not handled).
- Use square brackets (i.e. `[]`) rather than parentheses (i.e. `()`) to access vector entries. This is in line with how the STL accesses elements of containers.
- Renamed function parameters `rThisVariable` to `rVariable`.

avdg81 committed:
Renamed several function parameters

Also reformatted the code using Clang-tidy.

avdg81 committed:
Cleaned up (overrides of) `GetStrainSize`

- Moved the function implementations from the `.hpp` files to the `.cpp` files.
- Marked them `[[nodiscard]]`.
- Replaced any direct references to `VoigtSize` by calls to `GetStrainSize`. This change made the `static` members `VoigtSize` redundant (their values are now directly returned by `GetStrainSize`).

avdg81 committed:
Getting the `CAUCHY_STRESS_VECTOR` is more generic now

Consequently, one `override` that did the same thing has been removed, since it had become redundant.

avdg81 committed:
Used an STL algorithm to replace a raw `for` loop

avdg81 committed:
Removed a few overrides of `GetStrainMeasure`

The implementations were identical to what the base class does.

avdg81 committed:
Removed a few overrides of `GetStressMeasure`

The implementations were identical to what the base class does.

avdg81 committed:
Removed an override of `SetInternalStrainVector`

Its implementation was effectively the same as the one of the base class.

Other changes include:
- Removed a needless `KRATOS_TRY` and `KRATOS_CATCH` construct.
- Use the subscript operator `[]` to access vector elements rather than operator `()`.

avdg81 committed:
Removed an override of `CopyConstitutiveMatrix`

Its implementation was effectively the same as the one of the base class.

avdg81 committed:
Removed an override of `SetExternalStressVector`

Its implementation was effectively the same as the one of the base class.

avdg81 committed:
Removed an override of `SetValue`

Its implementation was effectively the same as the one of the base class.

avdg81 committed:
Cleaned up `Clone` members

- Removed needless `KRATOS_TRY` and `KRATOS_CATCH` constructs.
- Use `std::make_shared` rather than `Kratos::make_shared`.
- Marked them `[[nodiscard]]`.

avdg81 committed:
Cleaned up class `SmallStrainUDSM2DPlaneStrainLaw`

- All implementations are now in the `.cpp` file.
- Removed many redundant comments.
- Marked a few members `[[nodiscard]]`.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/clean-up-udsm-based-laws

avdg81 committed:
Cleaned up class `SmallStrainUDSM2DInterfaceLaw`

- All implementations are now in the `.cpp` file.
- Removed many redundant comments.
- Marked a member `[[nodiscard]]`.

avdg81 committed:
Cleaned up class `SmallStrainUDSM3DInterfaceLaw`

- All implementations are now in the `.cpp` file.
- Removed many redundant comments.
- Marked two members `[[nodiscard]]`.
- Corrected the base class used by the (de)serialization code.
- Removed a few redundant name qualifiers.

avdg81 committed:
Cleaned up class `SmallStrainUDSM3DLaw`

- All implementations are now in the `.cpp` file.
- Removed many redundant comments.
- Removed several redundant name qualifiers.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/clean-up-udsm-based-laws

avdg81 committed:
Further generalized class `SmallStrainUDSM3DLaw`

It now has a constitutive law dimension that keeps the strain size as well as the local dimension.

Attempted to improve the `Clone` member which did not clone all data members.

avdg81 committed:
Removed the derived class `SmallStrainUDSM2DPlaneStrainLaw`

Now that its base class (`SmallStrainUDSM3DLaw`) has a constitutive law dimension, the `override`s implemented by the derived class (`SmallStrainUDSM2DPlaneStrainLaw`) were no longer needed.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/clean-up-udsm-based-laws

avdg81 committed:
Several minor changes

- Moved the definition of a destructor to the implementation file. Hopefully, this will solve "error C2338: can't delete an incomplete type".
- Marked a constructor `explicit`.

avdg81 committed:
Added a comment

It references the source of the fixed array sizes.

avdg81 committed:
Extracted a member that clones data members

avdg81 committed:
`Clone` no longer uses the copy constructor

Instead, it uses the new member function `CloneDataMembersTo`. This is to prepare for the removal of the copy operations.

avdg81 committed:
Explicitly delete the copy operations and use defaulted move operations

avdg81 committed:
Added some `static_assert`s

These check at compile time that instances of the UDSM-based constitutive laws cannot be copied, but they can be moved.

avdg81 committed:
Renamed a class

avdg81 committed:
Several minor changes

- Removed a few redundant name qualifiers.
- Removed a few redundant comments.
- Corrected a few text strings.

avdg81 committed:
Added two missing `#include`s

This solves a compiler error.

avdg81 committed:
Removed several redundant name qualifiers

avdg81 committed:
Renamed two files

avdg81 committed:
Several minor changes

- Removed redundant usages of `this->`.
- Corrected a typo in a comment.
- `resize` unconditionally.
- Avoid needless aliases.

avdg81 committed:
Added a member that resets the constitutive matrix

avdg81 committed:
Changed two member functions into local helper functions

avdg81 committed:
Renamed some member functions

The new names comply with the Kratos Style Guide.

avdg81 committed:
Fixed a narrowing conversion

avdg81 committed:
Improved saving and loading of UDSM-based constitutive laws

avdg81 committed:
Removed a duplicated comment

avdg81 committed:
Converted an unscoped `enum` to a scoped one

avdg81 committed:
Removed four unused member functions

avdg81 committed:
Replaced an unscoped `enum` by a few locally defined indices

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_2D_interface_law.hpp:
```diff
@@ -12,215 +12,51 @@
 {
 public:
     // The base class ConstitutiveLaw type definition
     using BaseType = ConstitutiveLaw;
```
Its basetype is SmallStrainUDSMLaw not ConstitutiveLaw, but BaseType is not used in this file

avdg81 replied:
Good catch. I've removed it.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_2D_interface_law.hpp:
```diff
@@ -12,215 +12,51 @@
-    static constexpr SizeType VoigtSize = VOIGT_SIZE_2D_INTERFACE;
-
-    /// Pointer definition of SmallStrainUDSM2DInterfaceLaw
+    // Pointer definition of SmallStrainUDSM2DInterfaceLaw
```
Not useful comment, similar on 27

avdg81 replied:
Both have been removed.

avdg81 committed:
Changed the file extension of a few header files

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_2D_interface_law.h:
```diff
@@ -0,0 +1,62 @@
+    // The base class ConstitutiveLaw type definition
+    using BaseType = ConstitutiveLaw;
+
+    // The size type definition
```
BaseType not used here, and useless comments.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_2D_interface_law.cpp:
```diff
@@ -111,10 +127,29 @@ void SmallStrainUDSM2DInterfaceLaw::SetValue(const Variable<Vector>& rVariable,
+
+std::string SmallStrainUDSM2DInterfaceLaw::Info() const { return "SmallStrainUDSM2DInterfaceLaw"; }
+
+void SmallStrainUDSM2DInterfaceLaw::PrintInfo(std::ostream& rOStream) const { rOStream << Info(); }
```
Same as base class implementation.

avdg81 replied:
Good catch. I've removed it.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_3D_interface_law.cpp:
```diff
@@ -118,15 +130,34 @@ Vector& SmallStrainUDSM3DInterfaceLaw::GetValue(const Variable<Vector>& rVariabl
+
+std::string SmallStrainUDSM3DInterfaceLaw::Info() const { return "SmallStrainUDSM3DInterfaceLaw"; }
+
+void SmallStrainUDSM3DInterfaceLaw::PrintInfo(std::ostream& rOStream) const { rOStream << Info(); }
```
Same as base class implementation

avdg81 replied:
Removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_3D_interface_law.h:
```diff
@@ -0,0 +1,62 @@
+    // The base class ConstitutiveLaw type definition
+    using BaseType = ConstitutiveLaw;
+
+    // The size type definition
```
BaseType not used as far as I see, comments redundant.

avdg81 replied:
`BaseType` has been removed as well as the redundant comments.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -287,9 +257,18 @@ void SmallStrainUDSM3DLaw::ResetStateVariables(const Properties& rMaterialProper
-void SmallStrainUDSM3DLaw::ResetMaterial(const Properties&   rMaterialProperties,
-                                         const GeometryType& rElementGeometry,
-                                         const Vector&       rShapeFunctionsValues)
+void SmallStrainUDSMLaw::ResetConstitutiveMatrix()
```
Now defined once, thank you.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -588,83 +567,76 @@ void SmallStrainUDSM3DLaw::CalculateMaterialResponsePK1(ConstitutiveLaw::Paramet
+        << "The SmallStrainUDSMLaw needs an element provided strain" << std::endl;
 
     KRATOS_ERROR_IF(!rValues.IsSetStrainVector() || rValues.GetStrainVector().size() != GetStrainSize())
         << "Constitutive laws in the geomechanics application need a valid provided strain" << std::endl;
```
Out of scope probably, but it's nice if the Id and/or the name of the constitutive law are mentioned in the text to the user.

avdg81 replied:
Yes, I agree that at some point in time we should go over the error messages and improve them. But since that is not the aim of this PR, I will keep it the way it was for now.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.h:
```diff
@@ -0,0 +1,231 @@
+    void                      PrintData(std::ostream& rOStream) const override;
+
+protected:
+    enum class UdsmTaskId : int {
```
I wonder about the camel casing here. UDSM is mostly written completely in capitals.

avdg81 replied:
Using all caps makes it more difficult to see where the next word starts. I've read somewhere (I don't recall exactly where, sorry) that it might be better to apply upper camel case to acronyms, too. On the other hand, in our code we use all caps UDSM in more places. So for consistency's sake it may be better to use the all caps version. I have changed it.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -400,33 +378,34 @@ int SmallStrainUDSM3DLaw::GetNumberOfStateVariablesFromUDSM(const Properties& rM
     int    nStateVariables   = 0;
 
-    // variable to check if an error happend in the model:
+    // variable to check if an error occurred in the model:
```
UDSM constitutive law may be more descriptive than model.

avdg81 replied:
Or perhaps "the shared library" rather than "the UDSM constitutive law", since what the shared library (created by the end-user) does exactly is completely out of our control. What do you think?

WPK4FEM replied:
That is very correct:
variable to check if an error occurred in the shared library

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -673,39 +645,62 @@ void SmallStrainUDSM3DLaw::CopyConstitutiveMatrix(ConstitutiveLaw::Parameters& r
+void SmallStrainUDSMLaw::CalculateConstitutiveMatrix(Parameters& rValues, Matrix& rConstitutiveMatrix)
 {
     KRATOS_TRY
     // update strain vector
```
Redundant comment.

avdg81 replied:
Removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -673,39 +645,62 @@ void SmallStrainUDSM3DLaw::CopyConstitutiveMatrix(ConstitutiveLaw::Parameters& r
+void SmallStrainUDSMLaw::CalculateStress(Parameters& rValues, Vector& rStressVector)
 {
     KRATOS_TRY
     // update strain vector
```
Redundant comment.

avdg81 replied:
Removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -673,39 +645,62 @@ void SmallStrainUDSM3DLaw::CopyConstitutiveMatrix(ConstitutiveLaw::Parameters& r
 
+    auto task_id_as_int = static_cast<int>(TaskId);
+
     // process data
```
more process info than process data, the comment is redundant.

avdg81 replied:
Removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -737,186 +732,167 @@ void SmallStrainUDSM3DLaw::CallUDSM(int* pIDTask, ConstitutiveLaw::Parameters& r
-
-        rValue = 0.5 * inner_prod(r_strain_vector, r_stress_vector); // Strain energy = 0.5*E:C:E
+    if (rVariable == STRAIN_ENERGY) {
+        CalculateStress(rParameterValues, rParameterValues.GetStressVector());
+        rValue = 0.5 * inner_prod(rParameterValues.GetStrainVector(), rParameterValues.GetStressVector()); // Strain energy = 0.5*E:C:E
```
I'm tempted to place this implementation in the stress strain utilities, as I have seen this implementation in many places. 

The comment is wrong, its 0.5 eps : sigma here. The comment indicates a fully elastic model, which this is potentially not.

avdg81 replied:
Yes, I agree that would be a better place. However, in my opinion, this is beyond the scope of the current PR. We'd better create a new issue on the backlog to address that.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -737,186 +732,167 @@ void SmallStrainUDSM3DLaw::CallUDSM(int* pIDTask, ConstitutiveLaw::Parameters& r
         // Previous flags saved
-        const bool flagConstTensor = rFlags.Is(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
-        const bool flagStress      = rFlags.Is(ConstitutiveLaw::COMPUTE_STRESS);
+        const bool flagConstTensor = rFlags.Is(COMPUTE_CONSTITUTIVE_TENSOR);
+        const bool flagStress      = rFlags.Is(COMPUTE_STRESS);
```
Is it not easier to save and restore all flags at once i.s.o. toggle some of its components.

avdg81 replied:
I'd suggest to revisit this member function another time. What do you think?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -737,186 +732,167 @@ void SmallStrainUDSM3DLaw::CallUDSM(int* pIDTask, ConstitutiveLaw::Parameters& r
+        rFlags.Set(COMPUTE_CONSTITUTIVE_TENSOR, true);
+        rFlags.Set(COMPUTE_STRESS, true);
 
         // We compute the stress
```
What else.

avdg81 replied:
Removed the comment.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -737,186 +732,167 @@ void SmallStrainUDSM3DLaw::CallUDSM(int* pIDTask, ConstitutiveLaw::Parameters& r
 
     KRATOS_DEBUG_ERROR_IF(index < 0 || index > (static_cast<int>(mStateVariablesFinalized.size()) - 1))
-        << "GetValue: Variable: " << rThisVariable
+        << "GetValue: Variable: " << rVariable
         << " does not exist in UDSM. Requested index: " << index << std::endl;
```
Information about which UDSM or material Id would be nice.

avdg81 replied:
Agreed. But I'd suggest to make that part of a follow-up PR.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -927,16 +903,78 @@ void SmallStrainUDSM3DLaw::SetValue(const Variable<double>& rVariable, const dou
     } else if (rVariable == CAUCHY_STRESS_VECTOR) {
-        KRATOS_ERROR_IF(rValue.size() != StressVectorSize)
-            << "Failed to set stress vector: expected one with " << StressVectorSize
+        KRATOS_ERROR_IF(rValue.size() != GetStrainSize())
+            << "Failed to set stress vector: expected one with " << GetStrainSize()
             << " components, but got one with " << rValue.size() << "components\n";
```
A user has no clue which element, i.p., material Id etc.

avdg81 replied:
Also here I agree. However, I think it would be better to improve all the error messages for UDSM-based constitutive laws in a separate PR. What do you think?

WPK4FEM replied:
I agree.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.h:
```diff
@@ -0,0 +1,231 @@
+    static constexpr SizeType StressVectorSize          = 6;
+    static constexpr SizeType StrainIncrementVectorSize = 12;
+
+    /// Pointer definition of SmallStrainUDSMLaw
```
what else.

avdg81 replied:
Removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_2D_interface_law.cpp:
```diff
@@ -10,21 +10,41 @@
+    case INDEX_2D_INTERFACE_XZ:
+        return INDEX_3D_XZ;
+    default:
+        KRATOS_ERROR << "invalid index: " << Index2D << std::endl;
```
Not very informative to an end user.

avdg81 replied:
Let's create a separate issue on our backlog to improve the error messages.

WPK4FEM reviewed (commented):
Hi Anne,
Thank you for the massive clean up, functionality becomes clearer and more compact. My comments are what I see on a fast readthrough. Feel free to regards many of them as out of scope for this moment.
Regards, Wijtze Pieter

avdg81 committed:
Processed Wijtze Pieter's review suggestions

- Removed two redundant overrides (since they did the same thing as the base class implementation).
- Removed several redundant comments.
- Removed two unused type aliases.
- Changed `Udsm` to `UDSM`.

rfaasse commented on applications/GeoMechanicsApplication/geo_mechanics_application.h:
```diff
@@ -992,10 +991,10 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) KratosGeoMechanicsApplication : publ
-    const SmallStrainUDSM2DPlaneStrainLaw mSmallStrainUDSM2DPlaneStrainLaw{};
-    const SmallStrainUDSM2DInterfaceLaw   mSmallStrainUDSM2DInterfaceLaw{};
-    const SmallStrainUDSM3DInterfaceLaw   mSmallStrainUDSM3DInterfaceLaw{};
+    const SmallStrainUDSMLaw mSmallStrainUDSM2DPlaneStrainLaw{std::make_unique<PlaneStrain>()};
+    const SmallStrainUDSMLaw mSmallStrainUDSM3DLaw{std::make_unique<ThreeDimensional>()};
```
Very nice you were able to re-use the dimensions (and without any necessary changes to them) 🎉

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.h:
```diff
@@ -0,0 +1,231 @@
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Vahid Galavi
```
At the very least, I think you should add your own name here @avdg81 after this significant refactoring step

avdg81 replied:
No, I don't think my name should be there, since I wasn't one of the authors. Cleaning things up doesn't count as authorship in my opinion. But thanks for the appreciation. :-)

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -139,61 +131,25 @@ using f_UserMod          = void (*)(int*,
-    this->mStrainVectorFinalized   = rOther.mStrainVectorFinalized;
-    this->mSig0                    = rOther.mSig0;
+SmallStrainUDSMLaw::SmallStrainUDSMLaw(SmallStrainUDSMLaw&&) noexcept            = default;
+SmallStrainUDSMLaw& SmallStrainUDSMLaw::operator=(SmallStrainUDSMLaw&&) noexcept = default;
```
Would this not lead to issues with the mMatrixD pointer? I haven't worked too much with these kinds of c-style members, so I don't know (probably fine, but just checking)

avdg81 replied:
Oh no, I think you're entirely right. We'll probably suffer from array decay here. *&%^$ (that was the swearing part).

This definitely needs more thought. Since the move operations were previously disabled, I have disabled them again (but being explicit about it this time). Excellent catch! Thanks a lot.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -287,9 +257,18 @@ void SmallStrainUDSM3DLaw::ResetStateVariables(const Properties& rMaterialProper
-                                         const Vector&       rShapeFunctionsValues)
+void SmallStrainUDSMLaw::ResetConstitutiveMatrix()
+{
+    for (unsigned int i = 0; i < VOIGT_SIZE_3D; ++i) {
+        for (unsigned int j = 0; j < VOIGT_SIZE_3D; ++j) {
```
In other places VOIGT_SIZE_3D is replaced with GetStrainSize(). Should it be also here?
If there is need to have zeros in non-used elements, then would it be useful to use the following?
double mMatrixD[VOIGT_SIZE_3D][VOIGT_SIZE_3D]={{0.0}};

avdg81 replied:
Data member `mMatrixD` is a C-style multidimensional array with fixed sizes (`VOIGT_SIZE_3D` x `VOIGT_SIZE_3D`). For certain cases (e.g. plane strain), not all elements of this matrix will be used. Still, I feel it's a good habit to set all elements to zero (including the unused ones) when a reset is requested. For cases where you operate on the actual constitutive matrix (which may be a sub-matrix of `mMatrixD` as outlined above), we should use `GetStrainSize`. If you feel that I have overlooked a place where I should have used a different size, please let me know. Thanks!

Also, I have added `{}` to the declaration of `mMatrixD` to initialize it properly. I have used the syntax used in one of the examples from the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es20-always-initialize-an-object):
```cpp
constexpr int max = 8 * 1024;
int buf[max] = {};   // zero all elements; better in some situations
f.read(buf, max);
```

markelov208 replied:
Thank you for the explanation. Should {} be added to this matrix in SmallStrainUMAT3DLaw?

avdg81 replied:
I'm not sure what you mean. I have written the declaration as follows:
```cpp
double mMatrixD[VOIGT_SIZE_3D][VOIGT_SIZE_3D] = {}; // all elements equal zero
```

Does this answer your question? Or did you mean something else?

markelov208 replied:
Thank you. I think it is good to have this declaration in SmallStrainUDSMLaw. There is another mMatrix in SmallStrainUMAT3DLaw.

avdg81 replied:
Yes, it's there. I will leave class `SmallStrainUMAT3DLaw` untouched for now, to avoid further scope creep. Thank you for your suggestion.

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.h:
```diff
@@ -0,0 +1,231 @@
+    /// Pointer definition of SmallStrainUDSMLaw
+    KRATOS_CLASS_POINTER_DEFINITION(SmallStrainUDSMLaw);
+
+    explicit SmallStrainUDSMLaw(std::unique_ptr<ConstitutiveLawDimension> pDimension = nullptr);
```
Just for my information, this means that this constructor will be used in the serialization as the 'default' constructor right? Is that also how you intended it?

avdg81 replied:
Yes, what you wrote is correct and this is what I intended, too.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/small_strain_udsm_law.cpp:
```diff
@@ -673,40 +645,62 @@ void SmallStrainUDSM3DLaw::CopyConstitutiveMatrix(ConstitutiveLaw::Parameters& r
+    rDestination.mStressVector          = mStressVector;
+    rDestination.mDeltaStrainVector     = mDeltaStrainVector;
+    rDestination.mStrainVectorFinalized = mStrainVectorFinalized;
+    for (unsigned int i = 0; i < VOIGT_SIZE_3D; ++i) {
+        for (unsigned int j = 0; j < VOIGT_SIZE_3D; ++j) {
```
If VOGT_SIZE_3D in ResetConstitutiveMatrix() is replaced with GetStrainSize(), then it should be done here as well.

avdg81 replied:
Since we are cloning the elements of `mMatrixD`, I feel it's safest to copy them all, even any unused elements. The point is that I'd like to avoid having uninitialized memory. If you feel that's not needed here, could you please explain the rationale behind it? Thanks.

markelov208 replied:
Hi Anne, I though that mMatrixD is initialized with zeroths at the object creation stage. Then  there is  no need to reset/copy zeroths. Obviously, your version is safer.

rfaasse reviewed (commented):
Thanks @avdg81 for taking the time to make some major improvements in this area. I didn't get enough time to go through everything in detail, so I mainly have some more conceptual questions, but this is a very nice clean-up!

markelov208 reviewed (dismissed):
Hi Anne, that is a very impressive reduction of lines. I made comments on multidimensional array. Hope it is useful.

WPK4FEM reviewed (dismissed):
Good to go for me.

avdg81 committed:
Deleted the move operations

The default semantics of the move operations doesn't properly handle C-style multidimensional arrays (here `mMatrixD`). Previously, the move operations were deleted, too (since only the copy operations were implemented). When we have a good idea about how to handle the C-style multidimensional array, we can try to make the constitutive laws moveable.

Also, `mMatrixD` is now set properly initialized to a zero matrix.

avdg81 committed:
Improved one comment and removed another one

markelov208 reviewed (approved):
Hi Anne, I think it is good to merge.

WPK4FEM reviewed (approved):
You did a lot here, lets merge and do continuations later and separately.
